src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -0,0 +1,395 @@
+        logging.info("###########################################################")
+
+        # ======= State kept for use by all tests =======
+
+        # All endpoints in "full object" indexing format
+        self.endpoints = wildcard_read.attributes
+
+        # All endpoints in raw TLV format
+        self.endpoints_tlv = wildcard_read.tlvAttributes
+
+    def get_test_name(self) -> str:
+        """Return the function name of the caller. Used to create logging entries."""
+        return sys._getframe().f_back.f_code.co_name
+
+    def fail_current_test(self, msg: Optional[str] = None):
+        if not msg:
+            # Without a message, just log the last problem seen
+            asserts.fail(msg=self.problems[-1].problem)
+        else:
+            asserts.fail(msg)
+
+    # ======= START OF ACTUAL TESTS =======
+    def test_endpoint_zero_present(self):
+        logging.info("Validating that the Root Node endpoint is present (EP0)")
+        if 0 not in self.endpoints:
+            self.record_error(self.get_test_name(), location=AttributePathLocation(endpoint_id=0),
+                              problem="Did not find Endpoint 0.", spec_location="Endpoint Composition")
+            self.fail_current_test()
+
+    def test_descriptor_present_on_each_endpoint(self):
+        logging.info("Validating each endpoint has a descriptor cluster")
+
+        success = True
+        for endpoint_id, endpoint in self.endpoints.items():
+            has_descriptor = (Clusters.Descriptor in endpoint)
+            logging.info(f"Checking descriptor on Endpoint {endpoint_id}: {'found' if has_descriptor else 'not_found'}")
+            if not has_descriptor:
+                self.record_error(self.get_test_name(), location=AttributePathLocation(endpoint_id=endpoint_id, cluster_id=Clusters.Descriptor.id),
+                                  problem=f"Did not find a descriptor on endpoint {endpoint_id}", spec_location="Base Cluster Requirements for Matter")
+                success = False
+
+        if not success:
+            self.fail_current_test("At least one endpoint was missing the descriptor cluster.")
+
+    def test_global_attributes_present_on_each_cluster(self):
+        logging.info("Validating each cluster has the mandatory global attributes")
+
+        @dataclass
+        class RequiredMandatoryAttribute:
+            id: int
+            name: str
+            validator: Callable
+
+        ATTRIBUTE_LIST_ID = 0xFFFB
+
+        ATTRIBUTES_TO_CHECK = [
+            RequiredMandatoryAttribute(id=0xFFFD, name="ClusterRevision", validator=check_int_in_range(1, 0xFFFF)),
+            RequiredMandatoryAttribute(id=0xFFFC, name="FeatureMap", validator=check_int_in_range(0, 0xFFFF_FFFF)),
+            RequiredMandatoryAttribute(id=0xFFFB, name="AttributeList",
+                                       validator=check_non_empty_list_of_ints_in_range(0, 0xFFFF_FFFF)),
+            # TODO: Check for EventList
```

Comment:
This is implemented now, so in theory it should be available for 1.2 devices. Is there a way to tell what matter version a device is certified to? We must have rev'd the descriptor version, right? 

Not a blocker for now, just something to consider for later impl. Maybe a test that the CD matches to the correct cluster revisions is something to consider for cert.

Reply:
Will do in follow-up